File: Niveau 4/Niveau 4.0/camping.py
#-------------------------------------------------------------------------------
# Name:        Installation du camping
# Purpose:     France-IOI niveau 4.0
#
# Author:      Sébastien
#
# Created:     11/11/2022
#-------------------------------------------------------------------------------

import time
import sys
input = sys.stdin.readline

def main():
    l, c = map(int, input().split())
    plan = [list(map(lambda v : v == "1", input().split())) for _ in range(l)]
    carres = [[0]*c for _ in range(l)]
    taille_max = 0
    for y in range(l):
        for x in range(c):
            if plan[y][x]:
                carres[y][x] = 0
            elif y == 0 or x == 0:
                carres[y][x] = 1
            else:
                carres[y][x] = 1 + min(carres[y-1][x], carres[y][x-1], carres[y-1][x-1])
            taille_max = max(taille_max, carres[y][x])
    print(taille_max)


if __name__ == '__main__':
    start = time.time()
    main()
    end = time.time()
    print("Temps d'exécution :", end-start, "ms")


"""
6 7
1 0 0 1 0 0 1
0 0 0 0 0 0 0
1 0 0 0 0 0 0
0 0 0 0 0 0 0
0 1 0 0 0 0 1
1 0 0 0 1 0 1
---
5 7
0 0 0 1 1 1 1
0 0 0 1 1 1 1
0 0 0 1 1 1 1
1 1 1 1 1 0 0
1 1 1 1 1 0 0
---
4 5
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
"""

File: Niveau 4/Niveau 4.0/camping.hs
--------------------------------------------------------------------------------
-- Name:        Installation du camping
-- Purpose:     France-IOI niveau 4.0
--
-- Author:      Sébastien
--
-- Created:     11/11/2022
--------------------------------------------------------------------------------

import Control.Monad
import Data.List
import Data.Array.Unboxed
import System.TimeIt

type Coord = (Int, Int)
type Plan = Array Coord Int

main :: IO ()
main = timeIt (do
    (l:c:_) <- fmap (map read . words) getLine :: IO [Int]
    plan <- fmap (listArray ((1,1), (l,c)) . map read . words) getContents :: IO Plan
    print $ maximum (carres plan))

carres :: Plan -> Array Coord Int
carres plan =
    let (_, (l,c)) = bounds plan
        a = array ((1,1), (l,c))
                ([((1,x), 1-plan!(1,x)) | x <- [1..c]] ++
                 [((y,1), 1-plan!(y,1)) | y <- [2..l]] ++
                 [((y,x), if plan!(y,x) == 1 
                          then 0
                          else 1 + (minimum [a!(y,x-1), a!(y-1,x-1), a!(y-1,x)]))
                                        | y <- [2..l], x <- [2..c]])
    in a
